src/TrackTraversingAlgorithms.cpp

```diff
@@ -4,6 +4,147 @@
 
 
 /**
+ * @brief Constructor for MaxOpticalLength calls the TraverseTracks
+ *        constructor and sets the max optical path length to zero
+ * @param track_generator The TrackGenerator to pull tracking information from
+ */
+MaxOpticalLength::MaxOpticalLength(TrackGenerator* track_generator)
+                                 : TraverseTracks(track_generator) {
+  _max_tau = 0;
+}
+
+
+/**
+ * @brief Determines the maximum optical path length for the TrackGenerator
+ *        provided during construction
+ * @details The maximum optical path length is initialized to infinity for
+ *          segmentation within the TrackGenerator. Then Tracks are traversed
+ *          and onTrack(...) is applied, calculating a maximum optical length
+ *          and setting it on the TrackGenerator.
+*/
+void MaxOpticalLength::execute() {
+  FP_PRECISION infinity = std::numeric_limits<FP_PRECISION>::max();
+  _track_generator->setMaxOpticalLength(infinity);
+#pragma omp parallel
+  {
+    loopOverTracks(NULL);
+  }
+  _track_generator->setMaxOpticalLength(_max_tau);
+}
+
+
+/**
+ * @brief Calculates the optical path length for the provided segments and
+ *        updates the maximum optical path length if necessary
+ * @param track The track associated with the segments
+ * @param segments The segments for which the optical path length is calculated
+ */
+void MaxOpticalLength::onTrack(Track* track, segment* segments) {
+  for (int s=0; s < track->getNumSegments(); s++) {
+    FP_PRECISION length = segments[s]._length;
+    Material* material = segments[s]._material;
+    FP_PRECISION* sigma_t = material->getSigmaT();
+
+    for (int e=0; e < material->getNumEnergyGroups(); e++) {
+      FP_PRECISION tau = length*sigma_t[e];
+      if (tau > _max_tau) {
+#pragma omp critical
+        _max_tau = std::max(_max_tau, tau);
+      }
+    }
+  }
+}
+
+
+/**
+ * @brief Constructor for SegmentSplitter calls the TraverseTracks
+ *        constructor
+ * @param track_generator The TrackGenerator to pull tracking information from
+ */
+SegmentSplitter::SegmentSplitter(TrackGenerator* track_generator)
+                               : TraverseTracks(track_generator) {
+}
+
+
+/**
+ * @brief Splits segments stored explicity along each Track
+ * @details No MOCKernels are initialized for this function.
+ */
+void SegmentSplitter::execute() {
+#pragma omp parallel
+  {
+    loopOverTracks(NULL);
+  }
+}
+
+
+/**
+ * @brief Segments for the provided Track are split so that no segment has a
+ *        larger optical path length than the maximum optical path length
+ * @param track The Track whose segments are potentially split
+ * @param segments The segments associated with the Track
+ */
+void SegmentSplitter::onTrack(Track* track, segment* segments) {
```

Comment:
Is it necessary to have an argument for the segments? The track should have an array of its segments.